pkg/vm.go

```diff
@@ -0,0 +1,113 @@
+package pkg
+
+import (
+	"fmt"
+	"net"
+	"os"
+)
+
+//go:generate zbusc -module vmd -version 0.0.1 -name manager -package stubs github.com/threefoldtech/zos/pkg+VMModule stubs/vmd_stub.go
+
+// VMNetworkInfo structure
+type VMNetworkInfo struct {
+	// Namespace name
+	Namespace string
+	// Tap device name
+	Tap string
+	// Mac address of the device
+	MAC net.HardwareAddr
+	// Address of the device in the form of cidr
+	AddressCIDR string
+	// Gateway gateway address
+	GatewayIP string
+	// Nameservers dns servers
+	Nameservers []string
+}
+
+// Disk specifies vm disk params
+type Disk struct {
+	// Size is disk size in Mib
+	Size uint64
+}
+
+// VM config structure
+type VM struct {
+	// virtual machine name, or ID
+	Name string
+	// CPU is number of cores assigned to the VM
+	CPU uint8
+	// Memory size in Mib
+	Memory int64
+	// Network is network info
+	Network VMNetworkInfo
+	// an allocated storage for the vm operation
+	// where files/virtual disks can be allocated
+	Storage string
+	// KernelImage path to uncompressed linux kernel ELF
+	KernelImage string
+	// KernelArgs to override the default kernel arguments. (default: "ro console=ttyS0 noapic reboot=k panic=1 pci=off nomodules")
+	KernelArgs string
+	// RootImage path to root disk image
+	RootImage string
+	// Disks are a list of disks that are going to
+	// be auto allocated on the provided storage path
+	Disks []Disk
+}
+
+// Validate vm data
+func (vm *VM) Validate() error {
+	missing := func(s string) bool {
+		return len(s) == 0
+	}
+
+	if missing(vm.Name) {
+		return fmt.Errorf("name is required")
+	}
+
+	if missing(vm.KernelImage) {
+		return fmt.Errorf("kernel-image is required")
+	}
+
+	if missing(vm.RootImage) {
+		return fmt.Errorf("root-image is required")
+	}
+
+	if missing(vm.Storage) {
+		return fmt.Errorf("storage is required")
+	}
+
+	if vm.Memory <= 0 {
```

Comment:
We should also have a minimum. What is the required amount of memory for the VM to boot ?

Reply:
Let's take the k3os image as reference for the moment since this is going to be the first use case.